### PR TITLE
support adding arbitrary file path + refactor modals

### PR DESCRIPTION
**Pull request by danielzsh**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._
Adds an option to create a new solution from an arbitrary path:
1. 
image
2. 
image
3. profit
- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

### CODE REVIEW

**Comment by SansPapyrus683** [comment]
really appreciate this editor stuff
but could you possibly work on the issues here sometime :pleading_face:

**Comment by danielzsh** [comment]
After I finish this I promise 🙏🙏

File: src/components/Editor/EditorFileModalInterface.tsx
Comment (thecodingwizard) [inline src/components/Editor/EditorFileModalInterface.tsx:56]:
```suggestion
  const { hits } = useHits() as { hits: AlgoliaEditorFileHit[] };
  if (!hits.find(hit => hit.id === 'none')) {
```

File: src/components/Editor/EditorFileModalInterface.tsx
Comment (thecodingwizard) [inline src/components/Editor/EditorFileModalInterface.tsx:111]: ?

File: src/components/Editor/AddProblemModal.tsx
Comment (thecodingwizard) [inline src/components/Editor/AddProblemModal.tsx:25]: this modal is also used for the "Add Problem" button in `module.problems.json`; in that use case, I think we do want the solutionMetadata to be autofilled to whatever the source-specific solution metadata is.

**Review by thecodingwizard — COMMENTED** [review]
nice!! what do you think of also adding a "Add New Problem (Solution)" button on the sidebar, similar to the "Login with Github" button?

It also seems like it might be nice to support adding internal solutions for generic problems (ie. problems that don't have a parser). we could do this in a separate PR though.

there's also a bunch of unnecessary console.log's...

File: src/components/Editor/AddFileModal.tsx
Comment (thecodingwizard) [inline src/components/Editor/AddFileModal.tsx:56]: I feel like this code is duplicated in multiple places in the codebase (https://github.com/cpinitiative/usaco-guide/blob/bbc8e5b44a4eb3f6f080672b02f76a950801045a/src/components/ProblemSuggestionModal.tsx#L250), (https://github.com/cpinitiative/usaco-guide/blob/bbc8e5b44a4eb3f6f080672b02f76a950801045a/src/components/ProblemsPage/Selection.tsx#L73), maybe more. we may want to clean it up sometime

Comment (thecodingwizard) [reply src/components/Editor/AddFileModal.tsx:56]: it's not immediately related to this PR though, so feel free to do this in another PR if you don't want to do it now

File: src/components/Editor/EditorFileModalInterface.tsx
Comment (thecodingwizard) [inline src/components/Editor/EditorFileModalInterface.tsx:58]: Instead of doing this, maybe it's neater to just add another Button after `hits.map` for "Add New Problem".

File: src/components/QuizGeneratorModal.tsx
Comment (thecodingwizard) [inline src/components/QuizGeneratorModal.tsx:124]: not related to your PR, but for performance reasons I think it's better of `getCode()` was rewritten as `const code = useMemo(..., [quiz])` or something. probably not super important though

File: src/components/Editor/AddFileModal.tsx
Comment (danielzsh) [reply src/components/Editor/AddFileModal.tsx]: refactored!

File: src/components/Editor/EditorFileModalInterface.tsx
Comment (danielzsh) [reply src/components/Editor/EditorFileModalInterface.tsx:58]: That actually might be slightly more annoying, since styling will have to unnecessarily copied (if someone in the future wants to modify it, they'll have two things to modify...); gonna keep it like this for now!

File: src/components/Editor/AddProblemModal.tsx
Comment (danielzsh) [reply src/components/Editor/AddProblemModal.tsx:25]: sure! I'm not sure what to put for sources that aren't cf or usaco though (AC, CSES); any ideas?

**Comment by danielzsh** [comment]
Oh oops, just realized i never replied to this...
> nice!! what do you think of also adding a "Add New Problem (Solution)" button on the sidebar, similar to the "Login with Github" button?

I think I'm gonna keep it more discreet for now, since I believe usaco.guide does require approvals for adding new problems so really it should be a feature reserved for people who know what they're doing :)

File: src/components/Editor/EditorFileModalInterface.tsx
Comment (thecodingwizard) [reply src/components/Editor/EditorFileModalInterface.tsx:57]: Hmm the styling being unnecessarily copied is a good point, but I think in this case it's still better to have the duplication:

- The duplication is fairly short; it's not a large amount of duplication
- The duplication is right next to each other and it's obvious that it's duplicated, so making changes to both at the same time (ex. if we wanted to change some styles) is pretty fast (just ctrl+d to make multiple cursors on VSCode). Tailwind CSS wrote a blog post about this at some point (duplicated HTML classes was fine because of multiple cursors or something)
- Without the duplication, there's extra logic that's not immediately obvious. In particular, the onClick() handler now needs a separate branch for `if (hit.id === 'none')`, and it's not immediately obvious why that's there and what it's doing. With the duplication, the logic can be different and it's very clear why (one button is handling a normal hit and the other button is adding an option to create a new solution). We don't need any special checks.
- It's also strange that hits[] now has an extra item at the end of it. Imagine one day we wanted to make a "hits" counter in the popup -- we would have to remember to subtract 1 from the length of hits[].
- creating a new solution does not always have to have the same styling as the rest of the hits. For example, imagine we wanted to add a "plus" icon next to create new solution -- with the current approach, we would have to add another `hits.id === 'none'` check to conditionally render the icon. Or, we might want to make the new solution button have static position in the modal.

If you really didn't want to duplicate the button though, maybe you could change `hits.id` from `none` to `CREATE_NEW_SOLUTION` or something to make it super obvious what that's for? and / or change `// blank hit` to a comment explaining that this is a hit placeholder used to show a create new solution button. Or, you could refactor the button into a new component, but I think this is probably overkill.

(If you strongly want to keep it as is though, that's fine, I would just be curious as to why -- I haven't thought too much about this so maybe there's something I missed)

File: src/components/Editor/AddProblemModal.tsx
Comment (thecodingwizard) [reply src/components/Editor/AddProblemModal.tsx:25]: Maybe just "kind: none"? IDK what the current default is

Comment (thecodingwizard) [reply src/components/Editor/AddProblemModal.tsx:25]: Also see https://github.com/cpinitiative/usaco-guide/issues/904  and https://github.com/cpinitiative/usaco-guide/blob/master/src/models/problem.ts#L323; there's some discussion about auto-generated solutions

File: src/components/Editor/AddFileModal.tsx
Comment (thecodingwizard) [inline src/components/Editor/AddFileModal.tsx:17]:
```suggestion
  const createSol = useSetAtom(createNewInternalSolutionFileAtom);
  return (
```

File: src/components/Editor/AddFileModal.tsx
Comment (thecodingwizard) [inline src/components/Editor/AddFileModal.tsx:13]: it would be nice to add types to these

File: src/components/Editor/AddFileModal.tsx
Comment (thecodingwizard) [inline src/components/Editor/AddFileModal.tsx:26]: why ref and not controlled input (ie. `const [fileURL, setFileURL] = useState('')`?)

File: src/components/Editor/AddFileModal.tsx
Comment (thecodingwizard) [inline src/components/Editor/AddFileModal.tsx:46]: maybe add `disabled={fileStatus !== 'Create File'}` or something like that

File: src/components/Editor/AddProblemModal.tsx
Comment (thecodingwizard) [inline src/components/Editor/AddProblemModal.tsx:58]: same here -- why ref instead of `useState`?

File: src/components/Editor/AddProblemModal.tsx
Comment (thecodingwizard) [inline src/components/Editor/AddProblemModal.tsx:68]: should probably add `disabled={status!=...}` here too, otherwise if someone clicks the button twice `addProblem` might get run twice

File: src/components/Editor/AddProblemModal.tsx
Comment (thecodingwizard) [inline src/components/Editor/AddProblemModal.tsx:80]: nice animation :)

File: src/components/Editor/parsers/cses.ts
Comment (thecodingwizard) [inline src/components/Editor/parsers/cses.ts:10]: i think there's a default CF solution metadata

File: src/components/Editor/parsers/parse.ts
Comment (thecodingwizard) [inline src/components/Editor/parsers/parse.ts:19]: I feel like the logic of parsing the HTML page belong on the server, not the client. It feels a bit weird to just have an `/api/fetch-html` endpoint like this. I'm fine with leaving it as-is though because it seems like a lot of effort to change it and I don't think it's a big deal

File: src/components/Modal.tsx
Comment (thecodingwizard) [inline src/components/Modal.tsx:13]: nice, thanks for refactoring this!

Comment (thecodingwizard) [reply src/components/Modal.tsx:13]: You might be interested in this: https://catalyst.tailwindui.com/

File: src/components/ProblemsPage/Selection.tsx
Comment (thecodingwizard) [inline src/components/ProblemsPage/Selection.tsx:8]: thanks for refactoring this!

File: src/components/QuizGeneratorModal.tsx
Comment (thecodingwizard) [inline src/components/QuizGeneratorModal.tsx:96]: I think this should be `useMemo`, not `useCallback`, and rather than calling `getCode` we should just say `const code = useMemo(...)`. Otherwise we'll end up calling the `getCode()` function every render anyway.

File: src/components/QuizGeneratorModal.tsx
Comment (thecodingwizard) [inline src/components/QuizGeneratorModal.tsx:143]: A key of idx is kind of sketchy (I think there may be some issues if you try deleting a quiz question in the middle of the list). The key should be unique per-question; one way to do this is to add a unique key to every question when you create it (ex. in `addQuestion()`, add an `id` property to the question that's `Math.random()` or `Date.now()`, and make that your key)

File: src/components/QuizGeneratorModal.tsx
Comment (thecodingwizard) [inline src/components/QuizGeneratorModal.tsx:166]: same here -- indexed key is sketchy especially when you are able to remove items

**Review by thecodingwizard — COMMENTED** [review]
looks good other than a few minor changes; thanks for all your work here!

File: src/components/Modal.tsx
Comment (danielzsh) [reply src/components/Modal.tsx:13]: Oh wow, it looks really nice...will definitely consider it moving forward! (not too willing to use it rn since its still in dev 😅 )

File: src/components/QuizGeneratorModal.tsx
Comment (danielzsh) [reply src/components/QuizGeneratorModal.tsx:96]: Ah right, mb!

File: src/components/Editor/AddFileModal.tsx
Comment (danielzsh) [reply src/components/Editor/AddFileModal.tsx:26]: Not sure, is there a significant difference? I always thought useRef on inputs would be faster, since typing wouldn't trigger a re-render.

File: src/components/Editor/EditorFileModalInterface.tsx
Comment (danielzsh) [reply src/components/Editor/EditorFileModalInterface.tsx:57]: icic, thanks for the detailed explanation; will be changing this!

File: src/components/Editor/AddFileModal.tsx
Comment (thecodingwizard) [reply src/components/Editor/AddFileModal.tsx:34]: I don't think the performance overhead will be that high... I think React generally prefers `useState`: https://stackoverflow.com/questions/76086163/what-is-better-usestate-or-useref-when-handling-multiple-form-inputs-in-react

Comment (danielzsh) [reply src/components/Editor/AddFileModal.tsx:34]: got it 👍🏻

**Comment by SansPapyrus683** [comment]
hey how's this going

**Comment by danielzsh** [comment]
pretty well, js need a few more fixes :)

File: src/components/Editor/parsers/cses.ts
Comment (danielzsh) [reply src/components/Editor/parsers/cses.ts:10]: isn't this the cses parser?

File: src/components/Editor/AddProblemModal.tsx
Comment (danielzsh) [reply src/components/Editor/AddProblemModal.tsx:32]: hm, none of these apply to the currently available parsers; will be putting kind: none for now!

**Review by thecodingwizard — APPROVED** [review]
lgtm! if you've tested it and you think it works, feel free to merge whenever you're ready.

Thanks for adding this feature!!

**Comment by SansPapyrus683** [comment]
💀

**Pull request merged**
